### ISSUE TITLE
Add @Child to complement @Children

### ISSUE THREAD

**Issue by acrane13** [unic/neba#158]
It was be great to add a `@Child` annotation which pulls back a single model based on path.

Right now I'm accomplishing this by either using the `@This Resource` to get the child model by path or cycling through all of the `@Children` to find the node I'm looking for and then using the `ResourceModelProvider` to convert the `Resource` into the right model type. It seems like a lot of boilerplate code that could be encapsulated in a `@Child` annotation.

**Comment by olaf-otto** [comment]
Hi @acrane13, thanks for reaching out! Could you provide a code example of what it is your are trying to achieve?

I think what you are doing might already be possible using the @Path annotation with a relative path, like so (note that I have also used the Optional annotation for a lazy-loading 1:1 relationship here):

```
@ResourceModel(types = "...")
public class TestModel {
    @Path("child/path")
    private Optional<ModelType> child;
}
```
Is this what you want to do?

**Comment by acrane13** [comment]
Hi @olaf-otto, thanks for the response. Here is an example of what I'm trying to do.

I have what I call an atomic component, in this case an image. This component has data fields, like alt text and description, and style fields, like css class. I've got the data and style data separated onto different tabs for the dialog and everything for that component works great.

Next I created what I call compound component, in this case a product summary. This component has a dialog with some fields that belong to it and it also includes the image component data tab. The summary component itself cq:includes the image component at a path of heroImage like this:

```
<%@page trimDirectiveWhitespaces="true" %><%
%><%@include file="/apps/testsite/global.jsp" %><%
%><neba:defineObjects /><%
slingRequest.setAttribute(ComponentContext.BYPASS_COMPONENT_HANDLING_ON_INCLUDE_ATTRIBUTE, true);
%><%
%><div class="l-grid">
  <div class="l-row">
    <div class="l-col">
      <cq:include path="heroImage" resourceType="testsite/components/atom/image"/>
    </div>
  </div>
</div><%
slingRequest.removeAttribute(ComponentContext.BYPASS_COMPONENT_HANDLING_ON_INCLUDE_ATTRIBUTE);
%>
```

Through some of the values set on the product summary dialog I need to override style values on the image component, so I created a product summary model that looks like this:

```
package com.testsite.aem.sling.model;

import com.testsite.aem.sling.model.atom.Image;
import io.neba.api.annotations.Path;
import io.neba.api.annotations.PostMapping;
import io.neba.api.annotations.ResourceModel;
import io.neba.api.annotations.This;
import lombok.Data;
import org.apache.commons.lang3.StringUtils;
import org.apache.sling.api.resource.Resource;

@Data
@ResourceModel(types = ProductSummary.COMPONENT)
public class ProductSummary extends PageBase {
  public static final String COMPONENT = "testsite/components/productSummary";

  @This
  Resource pageResource;

  String heroAlt;

  Image heroImage;

  @PostMapping
  public void initialize() {
    if (StringUtils.isNotEmpty(heroAlt)) {
      heroImage.setAlt(heroAlt);
    }
  }
}
```

With this model the heroImage always comes in null and the only way I could get access to that model is to add this to the top of the initialize method:

```
    Resource heroResource = pageResource.getChild("heroImage");
    if (heroResource != null) {
      BundleContext bundleContext = FrameworkUtil.getBundle(FundOverview.class).getBundleContext();
      ResourceModelProvider modelProvider = (ResourceModelProvider)bundleContext.getService(bundleContext.getServiceReference(ResourceModelProvider.class.getName()));
      heroImage = (Image)modelProvider.resolveMostSpecificModel(heroResource);
    }
```

The other way I've tried to do it is to grab all the `@Children` in the model and cycle through them one by one until I find the right path and then do the model provider lookup again.

I tried modifying the product summary model to include a `@Path` annotation but that doesn't work either. I've also tried it with and without the `@Reference` annotation as well.

This is a very simplistic example and most of the time the logic on what to change in the child objects can get complicated. If I was just setting the alt on a child component I know I could have done that through the dialog of the Product Summary.

Any help as to what I might be doing wrong would be greatly appreciated. Oh, this is all running on AEM 6.2 and Neba 4.1.0.

**Comment by olaf-otto** [comment]
Hi @acrane13, thanks for the details! From what I see, the ` Image heroImage` field should absolutely be populated with the adapted child resource, i.e. there should be no need for any post-processing.

I speculate that, maybe, there is some interference with lombok (NEBA is using reflection to determine fields to be injected) and I'll setup a test case for this based on the [neba sample project ](https://github.com/unic/neba-sample-project). Could you try two things in the mean time:

- Do you get a non-Null Image when you adapt imageResource directly in your `@PostMapping` method, like so:

```
    Resource heroResource = pageResource.getChild("heroImage");
    if (heroResource != null) {
      heroImage = heroResource.adaptTo(Image.class);
    }

```

- I assume you already tried this, but does annotating heroImage with this `@Path` annotation help (speculating that the field name is altered during compilation):

```
@Path("heroImage")
Image heroImage;

```
Thanks and regards,
Olaf

**Comment by acrane13** [comment]
Hi @olaf-otto, thanks for the response.

I had tried the adaptTo before but it wasn't returning the right model. I want/need the one from the cache so that when the child component renders it has the right information. The adaptTo was converting the values for that path on a model object, but that wasn't being forwarded on to the child so any changes I made were lost. I tried this a while ago, so I may not be remembering this correctly. I'll have to go and test it again to be sure.

I tried explicitly adding the `@Path` annotation to the heroImage and that didn't work either.

As for lombok(http://www.projectlombok.org/), from what I understand the annotations it provides are built out during compile time. There is no runtime library for lombok. The `@Data` annotation creates getters, setters, toString, hashCode, equals and required constructors for the class. They should be fully available to inspection.

Thanks for the help on this.
Adam

**Comment by olaf-otto** [comment]
Hi @acrane13 ,

I have reproduced your scenario in the neba sample project, see https://github.com/unic/neba-sample-project/tree/neba-158. 

There, I've created the following content structure:

```
  "productSummary": {
    "jcr:primaryType": "nt:unstructured",
    "sling:resourceType": "neba-sample/components/productSummary",
    "heroAlt":"Product summary hero alt text",
    "heroImage": {
      "jcr:primaryType": "nt:unstructured",
      "sling:resourceType": "neba-sample/components/image",
      "alt":"Original image alt text"
    }
  }
```

Along with a ProductSummary and Image model. The productSummary itself is included in the page.jsp.
Does this match your scenario?

Regarding your use of models and overriding properties: You are actually using a side-effect introduced by NEBA's request-scoped resource model cache. That is, if you load a model for a resource, you *may* get the same model of you subsequently adapt the same resource to the same model in the exact same way. That being said, don't use this implicit side effect. It depends on the cache state and the invocation order of models and creates an invisible dependency between the models.

Instead, you could explicitly add that use case into the Image model or pass the overridden via the views, e.g. using request attributes. For instance, if creatign a 1:1 relationship between Image and ProductSummary is an option, you could add a field of type

```
    @Path("..")
    private Optional<ProductSummary> summary;
```

to Image. Would that be an option?

**Comment by acrane13** [comment]
Hi @olaf-otto,

The scenario you put together roughly matches what I'm trying to do. Thanks. Does this return the model properly? If so, is there any way for you to test it in AEM?

The problem I'm running into is that the child could have any number of parents. I don't want the child to know about all of the potential parents, I just want it to know how to render itself. Then I can keep all the rules for that particular parent/child relationship on the parent and not clutter the child, which can exist without a parent as well.

As for the request scoped cache, this seems like a great feature to take advantage of instead of a side-effect. Because of the way I've structured my compound components I know the order in which the components will be rendered. Also, correct me if I'm wrong, but if I was getting a non-null object on the parent model for the child model, then wouldn't that have come from the cache as well? As far as the invisible dependency, I'm ok with that because my parent component explicitly includes the child so I'm already creating a relationship there.

Thanks for all the help on this.

**Comment by olaf-otto** [comment]
Hi @acrane13,

I have made an update to the sample project branch that makes it run in AEM 6.2. To do so:
- Install NEBA 4.1.0 via the package manager
- Install the bundles in the [thirdparties directory ](https://github.com/unic/neba-sample-project/tree/neba-158/thirdparties)
- Install the neba-sample-propject bundle and start it

Then, you should be able to request the sample project site via [http://localhost:4502/content/neba-sample.html](http://localhost:4502/content/neba-sample.html)

Regarding your usage of the cache, I don't recommend it because it is an opaque implementation and does not necessarily work the way you expect it to. For instance, as of NEBA 4.1.0, a model adapted in another model usually does not use the same cached instance as a model resolved in a view.

**Comment by acrane13** [comment]
Hi @olaf-otto ,

Thanks for your latest response. I ended up implementing your interface suggestion and things are working beautifully now. I just needed to change my mindset of the child reaching back up to it's parent instead of the parent forcing the information down. Plus I don't need to rely on the cache anymore because each time the child model is built it also builds the parent and asks for changes. Much cleaner and easier to read.

Thanks for all your help on this and pointing me in the right direction.
